os/net/netmgr/netdev_wifi.c

```diff
@@ -184,11 +251,74 @@ int trwifi_serialize_scaninfo(uint8_t **buffer, trwifi_scan_list_s *scan_list)
 
 int trwifi_post_event(struct netdev *dev, lwnl_cb_wifi evt, void *buffer, uint32_t buf_len)
 {
-	int res = _trwifi_handle_event(dev, evt, buffer, buf_len);
-	if (res < 0) {
-		// if network stack is not enabled. it needs to be restart
-		NET_LOGE(TAG, "critical error network stack is not enabled\n");
-		assert(0);
+	trwifi_evt *msg = (trwifi_evt *)kmm_zalloc(sizeof(trwifi_evt));
+	if (!msg) {
+		return -1;
+	}
+	msg->evt = evt;
+	msg->dev = dev;
+	if (buffer) {
+		msg->buf = (void *)kmm_malloc(buf_len);
+		if (!msg->buf) {
+			return -1;
+		}
+		memcpy(msg->buf, buffer, buf_len);
+		msg->buf_len = buf_len;
+	}
+	LOCK_EVTQUEUE(g_queue);
+	sq_addlast(&msg->entry, &g_queue.queue);
+	UNLOCK_EVTQUEUE(g_queue);
+	SIGNAL_EVTQUEUE(g_queue);
+	return 0;
+}
+
+void *_trwifi_event_handler(void *arg)
+{
+	while (1) {
+		WAIT_EVTQUEUE(g_queue);
+		LOCK_EVTQUEUE(g_queue);
+		trwifi_evt *evt = TRWIFI_GET_EVT(sq_remfirst(&g_queue.queue));
+		UNLOCK_EVTQUEUE(g_queue);
+		int res = _trwifi_handle_event(evt->dev, evt->evt, evt->buf, evt->buf_len);
+		if (res < 0) {
+			// if network stack is not enabled. it needs to be restart
+			NET_LOGE(TAG, "critical error network stack is not enabled\n");
+			assert(0);
+		}
+
+		res = lwnl_postmsg(LWNL_DEV_WIFI, evt->evt, evt->buf, evt->buf_len);
+		if (res < 0) {
+			NET_LOGE(TAG, "critical error network stack is not enabled\n");
+			assert(0);
+		}
+		if (evt->buf) {
+			kmm_free(evt->buf);
+			evt->buf_len = 0;
+		}
+		kmm_free(evt);
 	}
-	return lwnl_postmsg(LWNL_DEV_WIFI, (int)evt, buffer, buf_len);
+
+	NET_LOGE(TAG, "critical error\n", errno);
+	return NULL;
+}
+
+int trwifi_run_handler(void)
+{
+	if (sem_init(&g_queue.lock, 0, 1) != 0) {
+		NET_LOGE(TAG, "init lock fail %d\n", errno);
+		return -1;
+	}
+	if (sem_init(&g_queue.signal, 0, 0) != 0) {
+		NET_LOGE(TAG, "init signal fail %d\n", errno);
+		return -1;
```

Comment:
same as above.

Reply:
I think ERROR is used to interact to system. because it's defined in sys/types.h. So in this case, returning -1 is right because it's only called by netmgr inside.